### ISSUE TITLE
Window management is poor for stand-alone plugins

### ISSUE THREAD

**Issue by v-fox** [sadko4u/lsp-plugins#73]
Using more than 1 stand-alone LSP plugins on one desktop is not an easy task because they:
* always start at leftmost screen even when another display is the primary;
* do not save their position;
* do not have "class" and "role" defined which a window manager could have used as hints for creating rules to force their position, such as moving to another real or virtual screen;
* do not scale their size manually, in addition to not auto-scaling with DPI as per #68.

**Comment by sadko4u** [comment]
> always start at leftmost screen even when another display is the primary
Have had not enough resources to test it. In two-monitor configuration, plugins always have started on the main monitor. Also, it's strange that plugins always start on the leftmost top position because plugins always start at the middle of the screen in my environment. The window manager may override these settings in forced mode. Which window manager/desktop environment are you currently using?

> do not save their position
Consider the case in which you've started two instances of the same plugin. How should they remember their positions?

> do not have "class" and "role" defined
Please explain more detailed in terms of X11 protocol. What does mean 'class' and 'role'? LSP Plugins do set window type:
https://github.com/sadko4u/lsp-plugins/blob/6a598435b35e9faf10de621cc40c460021c4eccc/src/ui/ws/x11/X11Window.cpp#L438-L573

> do not scale their size manually
Currently it's a restriction due to the case that not all widgets are scaling properly. This is a long-time job to review all widgets and make them scaling properly.

**Comment by v-fox** [comment]
> > always start at leftmost screen even when another display is the primary
> 
> Have had not enough resources to test it. In two-monitor configuration, plugins always have started on the main monitor. Also, it's strange that plugins always start on the leftmost top position because plugins always start at the middle of the screen in my environment. The window manager may override these settings in forced mode. Which window manager/desktop environment are you currently using?

Weirdly, they all start on right position of the left monitor for me, stacked on each other. The problem is that I keep it switched off unless needed.

> > do not save their position
> 
> Consider the case in which you've started two instances of the same plugin. How should they remember their positions?

As they start on each other anyway, they could do it both in the right place. Or on the right virtual desktop & monitor, at least. Normally apps would show windows on "the active screen" which is Window Manager's viewport (a virtual desktop on a real monitor) where mouse cursor has focus currently and then WM would decide their exact placement. But LSP plugins do not want to do it for some reason.

In addition to that, you might have seen how Calf has a system tray minimization feature. It's kind of clunky and sometimes fails (sometimes on restore host window is drawn empty and all controls are inaccessible). But is very helpful when running like 5 things that are not needed to be visible all the time and may eat up more CPU for graphics (too bad there is no [OpenCL offloading](https://github.com/GPUOpen-LibrariesAndSDKs/TAN)) than audio processing.

> > do not have "class" and "role" defined
> 
> Please explain more detailed in terms of X11 protocol. What does mean 'class' and 'role'? LSP Plugins do set window type:
> 
> https://github.com/sadko4u/lsp-plugins/blob/6a598435b35e9faf10de621cc40c460021c4eccc/src/ui/ws/x11/X11Window.cpp#L438-L573

Not sure but neither `xprop` or `kwin_x11` detect 'class' and 'role' on LSP. While main window of Firefox, for example, gives this:
```
WM_CLASS(STRING) = "Navigator", "Nightly"
WM_WINDOW_ROLE(STRING) = "browser"
```
I guess it's something from this:
https://www.x.org/releases/X11R7.6/doc/xorg-docs/specs/ICCCM/icccm.html#client_support_for_session_management
https://tronche.com/gui/x/icccm/sec-4.html#WM_CLASS

> > do not scale their size manually
> 
> Currently it's a restriction due to the case that not all widgets are scaling properly. This is a long-time job to review all widgets and make them scaling properly.

I see. With so many knobs and buttons it would not be a trivial task to scale their size and maybe even positioning. But it may also be opportunity to make interface more "dynamic" in hiding and unhiding "secondary" controls.

Like compressor's settings and visualization for each band. although that is good as it is. The only annoying thing is manually changing those controls one-by-one if the same setting is needed for several or all bands but that's another matter entirely. …And another thing is that knobs & visualizers are limited to decibel ranges that are smaller than actual full 144db-per-24-bits audio range even for manual entry.

**Comment by sadko4u** [comment]
@v-fox 
I've created **github-issue-73** branch and pushed changes that are responsible for setting **WM_NAME**, **WM_CLASS** and **WM_WINDOW_ROLE** properties: 14b1f75ef3c5690c8fc6270333313649242c6ff4

Please check whether the situation is better now for you or not.

**Comment by v-fox** [comment]
> I've created **github-issue-73** branch and pushed changes that are responsible for setting **WM_NAME**, **WM_CLASS** and **WM_WINDOW_ROLE** properties: [14b1f75](https://github.com/sadko4u/lsp-plugins/commit/14b1f75ef3c5690c8fc6270333313649242c6ff4)
> 
> Please check whether the situation is better now for you or not.

The window IDs seem correct now, I think kwin can match them… it still doesn't want to move them to secondary virtual screen and minimize by default (otherwise visualization eats more CPU than actual audio processing which is why tray minimization is nice) but it's probably kwin's own issue. Maybe it wants a rule per process or doesn't apply them correctly, KDE guys like to break all their stuff on each release and not fix it for years or ever, after all.

**Comment by sadko4u** [comment]
Do you use KDE with X.Org or KDE with Wayland? I think it matters since I'm using KDE with X.Org.

**Comment by sadko4u** [comment]
> do not save their position
Also please tell more detailed about this. You mean when starting session or what?
For session management, there's yet another protocol that should be implemented. LSP toolkit currently doesn't support [https://en.wikipedia.org/wiki/X_session_manager](X Session Management) protocol.

**Comment by v-fox** [comment]
> Do you use KDE with X.Org or KDE with Wayland? I think it matters since I'm using KDE with X.Org.

> > do not save their position
> 
> Also please tell more detailed about this. You mean when starting session or what?
> For session management, there's yet another protocol that should be implemented. LSP toolkit currently doesn't support [X Session Management](https://en.wikipedia.org/wiki/X_session_manager) protocol.

Just plain old KDE5 on X. Kwin has window rules in `systemsettings5` -> "Window Management" -> "Window Rules". Previous build of LSP plugins always started on top of each other on the screen №1 (not the primary screen of the session but defined as "1" in kernel driver's GPU port enumeration) virtual desktop №1. I tried to force them to do it minimized on screen №2 virtual desktop №2 and keep positions that they were closed in by using "Apply Initially" and "Remember" directives in rules.

With these new changes when I edit the rules, kwin crashes and restarts itself. But now they are in correct positions (probably just because they didn't reset on crash instead of being actually applied) and on screen №2. But they still not minimized and not on virtual desktop №2. It's possible that this is the result of the crash and on full user session restart it will be fine, I have not tried new user session yet because I will not reboot the machine for a while.

It seem that "session management" implies saving the actual state of the application, not just basic window configuration, kwin should be able to do so for windows that it already controls… but, technically, it should not have had problems with unnamed windows either.

The last time I had entire user session to actually properly resume "as is" was KDE3 in 2008…

**Comment by sadko4u** [comment]
It's very strange behaviour, especially of KWin.
I have 64-bit openSUSE 15.1 on my desktop, 64-bit openSUSE Tumbleweed on another desktop and 32-bit openSUSE Tumbleweed in VirtualBox environment. And all just show plugins fine in KDE. By the other side, I haven't tried to make some rules for KWin.

**Comment by sadko4u** [comment]
Merged changes into **devel** branch: a1191b51bcf2d959672d8460ec1117fb06c92666

**Comment by sadko4u** [comment]
Partially available in 1.1.17 release.
The scaling option will be implemented within #68 issue
If there are other things that you still consider undone, please open another issue.
Closing this issue.
